File: sau/series.py
"""Reading a part number out of a filename.

A series arrives as a folder of files named `part1_<something>.mp4`, so the
episode order is already in the name and does not need to be retyped. Parsing
it is kept here, pure and separate from the database, because it is the one
piece of this feature that is worth testing exhaustively: get it wrong and
episodes publish out of order, which is the single failure a series cannot
absorb.
"""

from __future__ import annotations

import re
from collections.abc import Iterable
from dataclasses import dataclass
from pathlib import PurePosixPath

#: `part1_name`, `Part 02 - name`, `ep3_name`, `episode_4_name`.
#:
#: The separator after the digits is required. Without it `part12_x` and
#: `part1_2x` are the same string to a lazy matcher, and a series that
#: silently renumbers itself is worse than one that refuses to parse.
_PART_RE = re.compile(
    r"^\s*(?:part|ep|episode)[\s_\-.]*(\d{1,4})(?:[\s_\-.]+(?P<label>.*))?$",
    re.IGNORECASE,
)


class SeriesNameError(ValueError):
    """A filename does not carry a part number."""


@dataclass(frozen=True)
class ParsedPart:
    """What a series filename decomposes into."""

    index: int
    #: Whatever followed the part number, punctuation tidied. Free text: it is
    #: shown to the operator to confirm the right file matched, never published.
    label: str
    filename: str


def parse_part(filename: str) -> ParsedPart:
    """Pull the episode number out of a series filename.

    Raises `SeriesNameError` rather than guessing. A file that does not follow
    the convention is an operator mistake worth surfacing at upload time, not
    an asset to quietly assign a position to.
    """
    stem = PurePosixPath(filename).stem
    match = _PART_RE.match(stem)
    if match is None:
        raise SeriesNameError(
            f"{filename!r} does not start with a part number. "
            "Expected something like 'part1_name.mp4'."
        )

    index = int(match.group(1))
    if index < 1:
        # part0 is almost always a numbering slip, and it would sort ahead of
        # the real first episode.
        raise SeriesNameError(f"{filename!r} has part number {index}; parts start at 1.")

    label = (match.group("label") or "").strip()
    return ParsedPart(index=index, label=re.sub(r"[_\-.]+", " ", label).strip(), filename=filename)


def try_parse_part(filename: str) -> ParsedPart | None:
    """`parse_part`, but None instead of an exception. For bulk sorting."""
    try:
        return parse_part(filename)
    except SeriesNameError:
        return None


def missing_parts(indices: Iterable[int]) -> list[int]:
    """Episode numbers absent from an otherwise contiguous run, lowest first.

    A gap means a file was not uploaded, and the operator wants to know before
    the release schedule is built around what is there.
    """
    present = {i for i in indices if i >= 1}
    if not present:
        return []
    return [n for n in range(1, max(present)) if n not in present]


def duplicate_parts(indices: Iterable[int]) -> list[int]:
    """Episode numbers claimed more than once, lowest first."""
    seen: set[int] = set()
    repeats: set[int] = set()
    for index in indices:
        if index in seen:
            repeats.add(index)
        seen.add(index)
    return sorted(repeats)

File: console/src/domain/series.ts
/**
 * Series rules: reading an episode number out of a filename, and the
 * placeholders a caption template may use.
 *
 * The parser here mirrors `sau/series.py`. It exists so the operator sees the
 * episode number the moment a file is dropped, before anything is uploaded —
 * the server is still the authority, and it re-parses the name it is sent.
 * Two implementations of one rule is a cost worth paying only because the
 * alternative is a round trip per file just to render a list.
 *
 * Caption *rendering* is deliberately not mirrored: the preview comes from the
 * API, rendered by the same function the publish path uses, so what is shown
 * cannot drift from what publishes.
 */

import type { Job, Series, SeriesPart } from "../api/types";

/** `part1_name`, `Part 02 - name`, `ep3_name`, `episode_4_name`. */
const PART_RE = /^\s*(?:part|ep|episode)[\s_\-.]*(\d{1,4})(?:[\s_\-.]+(.*))?$/i;

export interface ParsedPart {
  index: number;
  /** Whatever followed the number. Shown to confirm the right file matched. */
  label: string;
}

function stem(filename: string): string {
  const base = filename.split("/").pop() ?? filename;
  const dot = base.lastIndexOf(".");
  return dot > 0 ? base.slice(0, dot) : base;
}

/**
 * Pull the episode number out of a filename, or null if it does not carry one.
 *
 * The separator after the digits is required, which is what stops `part12_x`
 * being read as part 1 — a series that silently renumbers itself is the worst
 * failure this feature has.
 */
export function parsePart(filename: string): ParsedPart | null {
  const match = PART_RE.exec(stem(filename));
  if (!match) return null;

  const index = Number.parseInt(match[1] ?? "", 10);
  // part0 would sort ahead of the real first episode.
  if (!Number.isFinite(index) || index < 1) return null;

  return { index, label: (match[2] ?? "").replace(/[_\-.]+/g, " ").trim() };
}

/** Episode numbers absent from an otherwise contiguous run, lowest first. */
export function missingParts(indices: readonly number[]): number[] {
  const present = new Set(indices.filter((n) => n >= 1));
  if (present.size === 0) return [];
  const highest = Math.max(...present);
  const gaps: number[] = [];
  for (let n = 1; n < highest; n += 1) if (!present.has(n)) gaps.push(n);
  return gaps;
}

/** Episode order — never the order the uploads happened to finish in. */
export function byPart<T extends { part_index: number }>(parts: readonly T[]): T[] {
  return [...parts].sort((a, b) => a.part_index - b.part_index);
}

/** A part is publishable until it has jobs; after that it is spoken for. */
export function isQueued(part: SeriesPart): boolean {
  return part.jobs.length > 0;
}

/** The jobs that gave up on this episode, one per platform that failed. */
export function failedJobs(part: SeriesPart): Job[] {
  return part.jobs.filter((job) => job.state === "failed");
}

/**
 * Whether anything is still riding on this episode.
 *
 * A part whose every job has failed holds nothing: the file was rejected and
 * no platform is going to act on it again unless it is retried. That is the
 * one case where replacing the episode is the right move, so it is not the
 * same question as `isQueued`, which governs whether publishing it again
 * would duplicate work.
 */
export function isLive(part: SeriesPart): boolean {
  return part.jobs.some((job) => job.state !== "failed");
}

export function pendingParts(series: Series): SeriesPart[] {
  return byPart(series.parts).filter((part) => !isQueued(part));
}

/**
 * The placeholders a template may use, with what each resolves to.
 *
 * Shown next to the template fields, because a placeholder an operator cannot
 * remember is one they will not use.
 */
export const TEMPLATE_FIELDS: readonly { name: string; description: string }[] = [
  { name: "{series}", description: "The series title, as it reads in the caption" },
  { name: "{series_en}", description: "English series title" },
  { name: "{part}", description: "This episode's number" },
  { name: "{total}", description: "Declared episode count, or how many are registered" },
  { name: "{next_part}", description: "The next episode's number; empty on the last one" },
  { name: "{next_teaser}", description: "The teaser line, rendered; empty on the last episode" },
  { name: "{hook}", description: "This episode's hook — the one line that varies" },
  { name: "{hashtags}", description: "The hashtag block for the platform being rendered" },
  { name: "{synopsis}", description: "The series synopsis" },
];

/** A one-line description of where a series stands, for the list. */
export function describeSeries(series: Series): string {
  const total = series.effective_total;
  const queued = series.parts.filter(isQueued).length;
  const gaps = series.missing_parts.length;

  const parts = [`${series.parts.length} of ${total || "?"} parts`];
  if (queued > 0) parts.push(`${queued} queued`);
  if (gaps > 0) parts.push(`${gaps} missing`);
  return parts.join(" · ");
}
